src/azul/terraform.py

```diff
@@ -721,6 +721,21 @@ def tf_config(self, app_name):
                 key.replace('.', '_'): value
                 for key, value in bucket_notifications.items()
             }
+            # To prevent a race condition by Terraform, we make the bucket
+            # notifications depend on the related aws_lambda_permission.
+            for notification in resources['aws_s3_bucket_notification'].values():
+                lambda_arns = [
+                    lambda_function['lambda_function_arn']
+                    for lambda_function in notification['lambda_function']
+                ]
+                permission_resources = []
+                for key, value in resources['aws_lambda_permission'].items():
+                    lambda_arn = value['function_name']
+                    if lambda_arn in lambda_arns:
+                        permission_resources.append(f'aws_lambda_permission.{key}')
```

Comment:
This can be a list comprehension.

Reply:
Then again, the smell here is that you're iterating repeatedly over the same set of `aws_lambda_permission` resources to find the ones that belong to the lambdas being notified in the current `aws_s3_bucket_notification` resource. This is the classic case of where building an index would make this more efficient and easier to understand. The index would be a dictionary mapping the function arn to a set of `aws_lambda_permission` objects.